### PR TITLE
[Traceable FSDP2] Add `_compiled_autograd_enabled` global state variable

### PR DESCRIPTION
**Pull request by yf225**
After https://github.com/pytorch/pytorch/pull/137821, we will no longer be able to call the Compiled Autograd state getter under Dynamo tracing. One solution is to cache the "Compiled Autograd enabled" state outside of compile for FSDP2, and just read from the cache when we need the check. This is implemented by this PR.

Fixes https://github.com/pytorch/pytorch/issues/138177.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138187

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @zhaojuanmao @mrshenli @rohan-varma @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @rec @xmfan

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 1]
```diff
@@ -153,6 +158,7 @@ def _lazy_init(self) -> None:
             raise RuntimeError(
                 f"FSDP requires a single root module but got {self._modules}"
             )
+        detect_compiled_autograd()
```
Comment (xmfan) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:161]: is this executing after we wrap the fsdp2 with torch.compile?

if so, we could use `torch._dynamo.config.compiled_autograd`

Comment (xmfan) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:161]: https://github.com/pytorch/pytorch/pull/138241

Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_state.py:161]: discussed in https://github.com/pytorch/pytorch/pull/138241 - we couldn't set `torch._dynamo.config.compiled_autograd=True` on automatically within the CA ctx manager, and thus won't be able to use it here

**Review by xmfan — APPROVED** [review]

**Review by awgu — APPROVED** [review]

**Comment by yf225** [comment]
Need to rebase on top of https://github.com/pytorch/pytorch/pull/138174 to ensure the unit test is actually run.

**Pull request closed without merging**